### PR TITLE
refactor to use dependancy injection / cleanup

### PR DESCRIPTION
**Pull request by dynamix**
https://trello.com/c/j8WJwgpd/3012-refactor-go-service-to-use-dependency-injection

The executor could not easily be extended with functionality not defined
in the go-service package. This PR changes this. Instead of using
direct dependancies, dependancy injection is introduced. Users of
go-service can create services executed by the executor which depend on
plain struct or services from other packages by using the service 
registry. This will make it easier to create other easily reusable service like for Slack, DM, Aerospike and other things. 

still needs improvement: 

- creating services for tests
- creating single shot services
- Init/Shutdown timeout handling

used in
- https://github.com/remerge/act/pull/323 
- EP
- https://github.com/remerge/dfc/pull/34/files / https://github.com/remerge/dfc/pull/36
- https://github.com/remerge/trigger/pull/37/files

### CODE REVIEW

**Comment by itszootime** [comment]
Additional thought: the "service" naming might be confusing here. We have a main service which is then composed of other services, except these sub-services don't share the responsibilities of the main service. Also consider when we add the service service to the dfc repository - we will then have two things called "service" (the _actual_ service and the dfc client) with totally different functions.

What about "component"?

File: debug_server.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+}
+
+func (s *debugServer) Init() error {
+	if err := s.Server.Init(); err != nil {
+		return err
+	}
+
+	s.serviceStartTime = time.Now()
+	go s.serveDebug()
```
Comment (gburanov) [inline debug_server.go]: imho this is not consistent that debugServer starts on init and server need ServeTLS to do it. It is kinda confusing. WDYT?

File: healthcheck.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+// temp to comply with service interface
+func (h *HealthChecker) Shutdown(os.Signal) {
+	h.Close()
+}
+
+// Run starts healthcheck loop.
+// This method can be safely called multiple times.
+func (h *HealthChecker) Run() {
```
Comment (gburanov) [inline healthcheck.go]: this in internal, right? It never used directly, just through Init()... so we could make it lower case

**Comment by gburanov** [comment]
Except several minor comments, the biggest problem is for me that the interface is not compatible with old implementation, so after we will merge this - we will need to change all implementations. But IMHO this was discussed.

Also, there are quite many conflicts, so I hope nothing will be lost after rebasing.

Also, IMHO this pr is too huge to find bugs, maybe in the future we can solve it by splitting  it into smaller prs, that are much easier to review, for example "switch to mkg, switch to go modules, etc)

Generally, @dynamix  already even created presentation about the new registry, so I think it is safe to merge it. Even if there will be bugs, they will be solved on the way.

File: .travis.yml [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
 cache:
   directories:
-  - vendor
+  - $GOPATH/pkg/mod
+  - $GOPATH/bin
+  - $HOME/.cache/go-build
 
 before_install:
-  - docker-compose up -d
+  - docker-compose up -d
```
Comment (gburanov) [inline .travis.yml]: how docker-compose will work? u remove the docker-compose.yml

File: base.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+}
+
+// HasOpenDebugForwardingConns checks if there are open connections  to debug listeners
+func (b *Base) HasOpenDebugForwardingConns() bool {
+	return b.debugForwader.hasOpenConnections()
+}
+
+func SafeCreate(ctor func(...interface{}) (interface{}, error), err error) interface{} {
+	if err != nil {
+		panic(err)
```
Comment (gburanov) [inline base.go]: as for me this is anti safe =)

Comment (gburanov) [reply base.go]: it is used only in Init now. What is the other use case for it? When should user call it?

File: main/example_service.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/remerge/go-service"
+
+
```
Comment (gburanov) [inline main/example_service.go]: why so many spaces?

**Comment by gburanov** [comment]
I approved, but maybe @itszootime  wants to check it as well...

File: base.go (hunk 4 above)
Comment (akaspin) [reply base.go]: The common naming for this kind of functions is `Must*`.

File: Makefile [hunk 6]
```diff
@@ -1,4 +1 @@
-PROJECT := go-service
-PACKAGE := github.com/remerge/$(PROJECT)
-
-include Makefile.common
+include mkf/Makefile.common
```
Comment (akaspin) [inline Makefile]: The better way is split this PR. For example migration to "mkf" can be done in separate PR.

**Comment by akaspin** [comment]
A few thoughts:

1. Massive file renames makes this PR very hard to review.
1. PR can be separated because some parts like build system is independent.
1. In best world APIs changes should be done in more soft way.
1. Maybe it's worth to split this repo to separate packages.

File: healthcheck.go (hunk 2 above)
Comment (dynamix) [reply healthcheck.go]: true 👍 , changed

File: base.go (hunk 4 above)
Comment (dynamix) [reply base.go]: true 👍 , changed @ `Must`

File: main/example_service.go (hunk 5 above)
Comment (dynamix) [reply main/example_service.go]: ups, removed

File: Makefile (hunk 6 above)
Comment (dynamix) [reply Makefile]: True @ mkf but I won't create another PR for a four letter change as the subtree is always the same in every repo.

File: .travis.yml (hunk 3 above)
Comment (dynamix) [reply .travis.yml]: 👍 removed

File: debug_server.go (hunk 1 above)
Comment (dynamix) [reply debug_server.go]: It is. But this is kind of intentional. The debug routes should be available as soon as possible. However accepting normal requests is a different story. Most of the time we only want to do that after an event (for example the snapshot was loaded). So it is the users job to call the `Serve` at the right point in time. We might change this behaviour at some point in the future - maybe when we implemented graceful restarts.

**Comment by dynamix** [comment]
> 1. Massive file renames makes this PR very hard to review.

There was almost no renaming. Some files and their content have been deleted and were 100% replaced. 

> 2. PR can be separated because some parts like build system is independent.

Agree. But not worth splitting it again.

> 3. In best world APIs changes should be done in more soft way.

Agree. Unfortunately the old API was just not worth keeping. 

> 4. Maybe it's worth to split this repo to separate packages.

What would the split be?

**Comment by dynamix** [comment]
@gburanov 

> Except several minor comments, the biggest problem is for me that the interface is not compatible with old implementation, so after we will merge this - we will need to change all implementations. But IMHO this was discussed.

Yep. This will cause some extra work but it should not be too much.

 
> Also, there are quite many conflicts, so I hope nothing will be lost after rebasing.

I merged it.
 
> Also, IMHO this pr is too huge to find bugs, maybe in the future we can solve it by splitting it into smaller prs, that are much easier to review, for example "switch to mkg, switch to go modules, etc)
> 
> Generally, @dynamix already even created presentation about the new registry, so I think it is safe to merge it. Even if there will be bugs, they will be solved on the way.

👍

**Pull request merged**